content_scripts/mode_insert.coffee

```diff
@@ -47,34 +47,61 @@ class InsertMode extends Mode
       else
         null
 
+    @boundShadowRoots = new WeakMap?()
+
     @push
       _name: "mode-#{@id}-focus"
       "blur": (event) => @alwaysContinueBubbling =>
-        target = event.target
+        target = event.path?[0] ? event.target
         # We can't rely on focus and blur events arriving in the expected order.  When the active element
         # changes, we might get "focus" before "blur".  We track the active element in @insertModeLock, and
         # exit only when that element blurs.
         @exit event, target if @insertModeLock and target == @insertModeLock
       "focus": (event) => @alwaysContinueBubbling =>
-        if @insertModeLock != event.target and DomUtils.isFocusable event.target
-          @activateOnElement event.target
-        else if event.target.shadowRoot
-          # A focusable element inside the shadow DOM might have been selected. If so, we can catch the focus
-          # event inside the shadow DOM. This fixes #853.
-          shadowRoot = event.target.shadowRoot
-          eventListeners = {}
-          for type in [ "focus", "blur" ]
-            eventListeners[type] = do (type) ->
-              (event) -> handlerStack.bubbleEvent type, event
-            shadowRoot.addEventListener type, eventListeners[type], true
-
-          handlerStack.push
-            _name: "shadow-DOM-input-mode"
-            blur: (event) ->
-              if event.target.shadowRoot == shadowRoot
-                handlerStack.remove()
-                for own type, listener of eventListeners
-                  shadowRoot.removeEventListener type, listener, true
+        # NOTE(mrmr1993): The first element of event.path gives us the element being focused, even when it is
+        # in an (open) shadow DOM.
+        target = event.path?[0] ? event.target
+        if ShadowRoot?
```

Comment:
Here `window.ShadowRoot?` is better: Chrome has `window.ShadowRoot` as early as version 36, but Firefox only declares a global `ShadowRoot` if `dom.webcomponents.enabled` is `true` in `about:config`.
BTW, now Vimium can not work if no `ShadowDOM` APIs, so in face this line is not needed.